agent/consul/leader_peering.go

```diff
@@ -27,8 +29,83 @@ import (
 	"github.com/hashicorp/consul/proto/pbpeerstream"
 )
 
+var leaderExportedServicesCountKey = []string{"consul", "peering", "exported_services"}
+var LeaderPeeringMetrics = []prometheus.GaugeDefinition{
+	{
+		Name: leaderExportedServicesCountKey,
+		Help: "A gauge that tracks how many services are exported for the peering. " +
+			"The labels are \"peering\" and, for enterprise, \"partition\". " +
+			"We emit this metric every 9 seconds",
+	},
+}
+
 func (s *Server) startPeeringStreamSync(ctx context.Context) {
 	s.leaderRoutineManager.Start(ctx, peeringStreamsRoutineName, s.runPeeringSync)
+	s.leaderRoutineManager.Start(ctx, peeringStreamsMetricsRoutineName, s.runPeeringMetrics)
+}
+
+func (s *Server) runPeeringMetrics(ctx context.Context) error {
+	ticker := time.NewTicker(s.config.MetricsReportingInterval)
+	defer ticker.Stop()
+
+	logger := s.logger.Named(logging.PeeringMetrics)
+	defaultMetrics := metrics.Default
+
+	retryLoopBackoff(ctx, func() error {
+		if err := s.emitPeeringMetrics(ctx, logger, ticker, defaultMetrics()); err != nil {
+			return err
+		}
+		return nil
+
+	}, func(err error) {
+		s.logger.Error("error emitting peering stream metrics", "error", err)
+	})
+
+	return nil
+}
+
+func (s *Server) emitPeeringMetrics(ctx context.Context, logger hclog.Logger, ticker *time.Ticker, metricsImpl *metrics.Metrics) error {
+	for {
+		select {
+		case <-ctx.Done():
+			logger.Info("stopping peering metrics")
+
+			// "Zero-out" the metric on exit so that when prometheus scrapes this
+			// metric from a non-leader, it does not get a stale value.
+			metrics.SetGauge(leaderExportedServicesCountKey, float32(0))
+			return nil
+		case <-ticker.C:
+			stateS := s.fsm.State()
+			ws := memdb.NewWatchSet()
+			ws.Add(stateS.AbandonCh())
+			ws.Add(ctx.Done())
```

Comment:
We're not blocking anything with this watchset; should it be removed?

Reply:
ah yes 🤦🏼 , thanks for spotting that!